pyanaconda/install_tasks.py

```diff
@@ -0,0 +1,400 @@
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+from threading import RLock
+from pyanaconda.isignal import Signal
+import time
+import logging
+log = logging.getLogger("anaconda")
+
+
+class TaskQueue(list):
+    """TaskQueue represents a queue of TaskQueues or Tasks.
+
+    TaskQueues and Tasks can be mixed in a single TaskQueue.
+    """
+
+    def __init__(self, name, status_message=None):
+        super(TaskQueue, self).__init__()
+        self._name = name
+        self._status_message = status_message
+        self._lock = RLock()
+        self._current_task_number = None
+        self._current_queue_number = None
+        self._running = False
+        self._done = False
+        self._parent = None
+        # triggered when this TaskQueue is started/completed
+        self.started = Signal()
+        self.completed = Signal()
+        # triggered if a TaskQueue contained in this one was started/completed
+        self.queue_started = Signal()
+        self.queue_completed = Signal()
+        # triggered when a task is started
+        self.task_started = Signal()
+        self.task_completed = Signal()
+
+        # connect to the task & queue started signals for
+        # progress reporting purposes
+        self.queue_started.connect(self._queue_started_cb)
+        self.task_started.connect(self._task_started_cb)
+
+    def _queue_started_cb(self, *args):
+        with self._lock:
+            self._current_queue_number += 1
+
+    def _task_started_cb(self, *args):
+        with self._lock:
+            self._current_task_number += 1
+
+    def __hash__(self):
+        """List are normally not hashable but our list subclass is.
+
+        Unlike with a plain list we care more about a queue instance than about it's content.
+        """
+        return id(self)
+
+    @property
+    def name(self):
+        """Task queue name.
+
+        :returns: task queue name
+        :rtype: str
+        """
+        return self._name
+
+    @property
+    def status_message(self):
+        return self._status_message
+
+    @property
+    def running(self):
+        """Reports if the task queue processing is currently running.
+
+        :returns: if the task queue is currently running
+        :rtype: bool
+        """
+        with self._lock:
```

Comment:
After investigating this a bit it doesn't seem that easy:
- AFAIK a lock instance can't be directly used as a decorator
- a custom decorator could be used as long **as the lock is named consistently** (eq. self._lock) - seem like quite a hack though :)

Based on [this](https://github.com/GrahamDumpleton/wrapt/blob/develop/blog/07-the-missing-synchronized-decorator.md#synchronizing-off-a-thread-mutex):

```
import functools
from threading import RLock

def synchronized(wrapped):
    @functools.wraps(wrapped)
    def _wrapper(*args, **kwargs):
        with args[0]._lock:
            return wrapped(*args, **kwargs)
    return _wrapper

class Test(object):
    def __init__(self):
        self._lock = RLock()

    @synchronized
    def function(self):
        print(self._lock)
        print("Hello world!")


test = Test()
test.function()
```

This actually works and is pretty lightweight, just a little crazy on the implementation side. If concerned parties are fine with this I can use this to replace the current locking usage on most places in the tasks PR. :)